src/errors.rs

```diff
@@ -0,0 +1,39 @@
+use std::fmt::{Display, Formatter};
+
+#[derive(Debug)]
+pub enum UrlParseError {
```

Comment:
You should create a type alias to `Result<T, UrlParseError>` and use it instead, e.g. in `src/utils.rs`' `check_uri` function.